### PR TITLE
feat: make our generalize trick work for arbitrary number of variables

### PR DESCRIPTION
**Pull request by tobiasgrosser**
This superseeds https://github.com/opencompl/ssa/pull/168.

### CODE REVIEW

File: SSA/Core/Tactic.lean [hunk 1]
```diff
@@ -67,31 +74,12 @@ macro "simp_peephole" "[" ts: Lean.Parser.Tactic.simpLemma,* "]" "at" ll:ident :
...
-      try clear c;
-      try clear b;
-      try clear a;
-      try revert f;
-      try revert e;
-      try revert d;
-      try revert c;
-      try revert b;
-      try revert a;
+      repeat (generalize_or_fail at $ll)
```
Comment (tobiasgrosser) [inline SSA/Core/Tactic.lean:82]: This already supports an arbitrary number of generalizations, but I am unsure how to revert an arbitrary number of variables. Anyone has an idea?

Comment (tobiasgrosser) [reply SSA/Core/Tactic.lean:82]: This now works as I figured out how to revert a variable of a given type.

**Review by alexkeizer — CHANGES_REQUESTED** [review]
Looking at line 304 (definition `p1`) of PaperExamples, the goal state post `simp-peephole` is
```
Γv : Ctxt.Valuation [Ty.int]
⊢ (fun v => v + v)^[0] (Γv { val := 0, property := ⋯ }) = Γv { val := 0, property := ⋯ }
```
whereas on the main branch, the goal is
```
⊢ ∀ (a : BitVec 32), (fun v => v + v)^[0] a = a
```

So it seems the refactor doesn't work 100% as intended, even if it doesn't break the build

File: SSA/Core/Tactic.lean [hunk 2]
```diff
@@ -49,6 +49,13 @@ private theorem Ctxt.destruct_nil {Ty} [TyDenote Ty] {f : Ctxt.Valuation ([] : C
   · intro h; apply h
   · intro h V; rw [Ctxt.Valuation.eq_nil V]; exact h
 
+local macro "generalize_or_fail" "at" ll:ident : tactic =>
+  `(tactic|
+      (
+        conv in ($ll { val := (_ : Nat), property := _ }) => skip;
```
[the remark this answers is not in the document]
Comment (tobiasgrosser) [reply SSA/Core/Tactic.lean:61]: I added a comment. Thank you for bringing this up.

File: SSA/Core/Tactic.lean [hunk 3]
```diff
@@ -73,26 +80,9 @@ macro "simp_peephole" "[" ts: Lean.Parser.Tactic.simpLemma,* "]" "at" ll:ident :
...
-      try clear a;
-      try revert f;
-      try revert e;
-      try revert d;
-      try revert c;
-      try revert b;
-      try revert a;
+      repeat (generalize_or_fail at $ll)
       try clear $ll;
+      repeat(revert ‹TyDenote.toType _›)
```
[the remark this answers is not in the document]
Comment (tobiasgrosser) [reply SSA/Core/Tactic.lean]: Great idea. I now revert immediately.

**Comment by tobiasgrosser** [comment]
> Looking at line 304 (definition `p1`) of PaperExamples, the goal state post `simp-peephole` is
> 
> ```
> Γv : Ctxt.Valuation [Ty.int]
> ⊢ (fun v => v + v)^[0] (Γv { val := 0, property := ⋯ }) = Γv { val := 0, property := ⋯ }
> ```
> 
> whereas on the main branch, the goal is
> 
> ```
> ⊢ ∀ (a : BitVec 32), (fun v => v + v)^[0] a = a
> ```
> 
> So it seems the refactor doesn't work 100% as intended, even if it doesn't break the build. We might want to remove the `try` in `try revert $ll`, so that such failures to remove the local context become errors (with a configuration to turn this behaviour of)

The problem is described here: https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/conv.20tactic.20sometimes.20solves.20goal/near/432918099

**Comment by alexkeizer** [comment]
> In brief the check if a pattern exits seems to solve trivial pattern and consequently prevents later leads to a failure of 'no goals to be solved' which then prevents the simplification we are after.

You could similarly guard the later tactics with a `first | done | ...`, that way won't prevent the closing of trivial goals, but it will prevent further 'no goals to be solved' errors. We already close some trivial goals with the `simp`-call, so I don't think that's so bad

**Comment by tobiasgrosser** [comment]
> ```
> first
>   | done
>   | simp (config := {failIfUnchanged := false}) only [Ctxt.Var.toSnoc, Ctxt.Var.last]
>      repeat (generalize_or_fail at $ll)
>      clear $ll
> ```

This breaks ScfFunctor.lean

**Comment by alexkeizer** [comment]
Ah, yeah it breaks `ScfFunctor` on main as well (I was just double checking my claim): we do apply the context, but only to a variable `v : Var ..` in the context, not an instance of `Var.mk _ _`, which is what we check for in generalize.

**Comment by tobiasgrosser** [comment]
> > In brief the check if a pattern exits seems to solve trivial pattern and consequently prevents later leads to a failure of 'no goals to be solved' which then prevents the simplification we are after.
> 
> You could similarly guard the later tactics with a `first | done | ...`, that way won't prevent the closing of trivial goals, but it will prevent further 'no goals to be solved' errors. We already close some trivial goals with the `simp`-call, so I don't think that's so bad

I dropped this for now. I feel this is the best we can do so far. Or would you like to see further changes.

Btw, lean3 had a tactic that failed if a child-tactic succeeded and the other way around. Do these still exist under a different name?

**Comment by alexkeizer** [comment]
I'm having a go at it now, we can just generalize the `generalize` line to work for all kinds of variables.
Since the `repeat (generalize_or_fail)` line may close the goal, we do need to add another layer of `first | done | ...` after it.

**Comment by alexkeizer** [comment]
So this change also breaks two proof in `AliveAutoGenerated`, where the context fails to clear.
Turns out, this failure also happens on main, we just don't notice there because the error is swallowed by `try`, and the rest of the proof is sorried out. IMO more evidence for why we *shouldn't* use `try`

**Comment by tobiasgrosser** [comment]
contains? is broken in our codebase:

https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/conv.20tactic.20sometimes.20solves.20goal/near/432918099

**Comment by tobiasgrosser** [comment]
> > So this change also breaks two proof in `AliveAutoGenerated`, where the context fails to clear. Turns out, this failure also happens on main, we just don't notice there because the error is swallowed by `try`, and the rest of the proof is sorried out. IMO more evidence for why we _shouldn't_ use `try`
> 
> Follow-up on this: the issue seems to be that the goal still mentions `Ctxt.Valuation.snoc \Gammav ... \<0, ...\>`. This _should_ reduce to `\Gammav \<0, ...\>`, but it isn't. I tried debugging it, but couldn't figure it out, and am not super keen to go too deep on this. Since it's already broken in `main` anyway, I've just moved the `sorry` up in the two affected theorems (`alive_Select_858` and `alive_Select_859`)

Thank you for investigating. We should get this fixed eventually, but I don't think we need to add this to this PR.

File: SSA/Core/Tactic.lean [hunk 4]
```diff
@@ -49,6 +51,30 @@ private theorem Ctxt.destruct_nil {Ty} [TyDenote Ty] {f : Ctxt.Valuation ([] : C
...
+  `(tactic|
+      (
+        -- We first check with `contains?` if the term is present in the goal.
+        -- This is needed as `generalize` never fails and just introduces a new
+        -- metavariable in case no variable is found. `contains?` will instead
+        -- fail if the term is not present in the goal.
+        contains? ($ll (_ : Var ..))
+        generalize ($ll (_ : Var ..)) = e at *;
+        simp (config := {failIfUnchanged := false}) only [TyDenote.toType] at e
+        revert e
```
[the remark this answers is not in the document]
Comment (tobiasgrosser) [reply SSA/Core/Tactic.lean:75]: > Does this still have the issue that it sometimes closes a trivial goal?
> 
> Some testing suggests it doesn't, so I've added a commit that replaces `first | done | clear $ll` with just `clear $ll` below
Thank you. This is great.

**Comment by tobiasgrosser** [comment]
I squashed your change. Is there anything else you want me to change, or can we merge this?

**Comment by alexkeizer** [comment]
I also removed the now-obsolete comment about needing `first | done | clear _`.
With this, I think it's good, feel free to squash and merge

**Comment by tobiasgrosser** [comment]
@alexkeizer, I think you need to approve this PR (and drop your merge-requested).

**Pull request merged**